Implement color conversion from RGB to YUV422 family

definitelyuncertain opened the pull request:
Related PR for extra: https://github.com/opencv/opencv_extra/pull/1104

Hi,

This patch provides CPU and OpenCL implementations of color conversions from RGB/BGR to YUV422 family (such as UYVY and YUY2).

These features would come in useful for enabling standard RGB images to be supplied as input to algorithms or networks that make use of images in YUV422 format directly (for example, on resource constrained devices working with camera images captured in YUV422).

The code, tests and perf tests are all written following the existing pattern. There is also an example `bin/example_cpp_cvtColor_RGB2YUV422` that loads an image from disk, converts it from BGR to UYVY and then back to BGR, and displays the result as a visual check that the conversion works.

The OpenCL performance for the forward conversion implemented here is the same as the existing backward conversion on my hardware. The CPU implementation, unfortunately, isn't very optimized as I am not yet familiar with the SIMD code.

Please let me know if I need to fix something or can make other modifications.

Thanks!

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
- [x] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Win64 OpenCL
```

opencv-alalek commented on samples/cpp/cvtColor_RGB2YUV422.cpp:
```diff
@@ -0,0 +1,72 @@
+#include "opencv2/core/utility.hpp"
```
> `samples/cpp/cvtColor_RGB2YUV422.cpp`

filenames should be lowercase

opencv-alalek commented on samples/cpp/cvtColor_RGB2YUV422.cpp:
```diff
@@ -0,0 +1,72 @@
+    string filename = parser.get<string>(0);
+
+    // Whether to use OpenCL
+    cv::ocl::setUseOpenCL(false);
```
This should not be used in samples (forcing mode).

opencv-alalek commented on samples/cpp/cvtColor_RGB2YUV422.cpp:
```diff
@@ -0,0 +1,72 @@
+        return -1;
+    }
+
+    image = orig_img.getUMat(ACCESS_READ);
```
Basic samples don't use `UMat` in general. Just `Mat` is enough.

`UMat` samples with OpenCL mode switching should go to `samples/tapi` location.

definitelyuncertain commented:
@opencv-alalek : Thanks for taking a look!

I've changed the sample filename to lowercase, and have removed the OpenCL parts from the sample code.

I'll be happy to work on a separate sample code in the `tapi`  folder that does the same thing with OpenCL if it'll be helpful.

Do let me know if you have more comments!

definitelyuncertain commented:
It looks like I misunderstood what test data means for performance tests.

I've sent in a [PR](https://github.com/opencv/opencv_extra/pull/1103) to opencv/extra with the updated sanity check data.

With this updated data, the perf tests are able to run successfully on my machine using the command from the presently unsuccessful CI run.

opencv-alalek commented:
Please take a look on issues from OpenCL build configuration: https://pullrequest.opencv.org/buildbot/builders/precommit_opencl/builds/100210

definitelyuncertain commented:
@opencv-alalek  Found the problem, it was a wrong step length that never executed on my machine due to the rows per OpenCL work item being set to 1, but did so on the build server.

definitelyuncertain commented:
@asmorkalov @opencv-alalek  I took a look at the errors from the recent run.

I've pushed a fix for the warning about the double incrementing that appeared in the MacOS builds.

For Ubuntu 22.04 and Android, it just says

```Error: tar: command not found```

during the initial setup. Not sure what that is, but let me know if there's anything I can do about it.

asmorkalov commented:
We have troubles with one of CI servers. Please ignore the issue right now. OpenCV team works on it.

asmorkalov commented:
I extended sample locally. Attached image is 30*absdiff(original, reconstructed):
![yuv_sample](https://github.com/opencv/opencv/assets/2536374/f808ce8d-9065-4cf8-b80f-b94f2cccab5b)

definitelyuncertain commented:
@asmorkalov @opencv-alalek Thanks for reviewing!

I cleaned up the code and made the sample code display the scaled diff as you did.

I've also mentioned my source for the coefficients where they appear and adjusted them to more rigorously derived values.

As for the results, I'm getting a visually similar scaled diff as above with the new coefficients as well:
![image](https://github.com/opencv/opencv/assets/20875718/f6161162-0fd0-4e76-94a2-b826c419b972)

IIUC, conversion from RGB to YUV422 isn't an invertible transformation on account of using 1 byte less per pixel, so some difference is unavoidable to begin with. Added to that, the values are clipped to 0-255 with finite precision, which is likely causing further losses.

Perhaps it'll work better with float images, but I'm not sure if anything can be done at all about integer images.

Let me know if you have any suggestions or other comments.

definitelyuncertain commented:
It looks like the `imgproc` tests are passing everywhere. Both the failures in Win64 and Win64 OpenCL seem to be in different modules: one is in test_umat_optical_flow from [test_python2-ippicv-opencl](https://pullrequest.opencv.org/buildbot/builders/precommit_opencl/builds/100215/steps/test_python2-ippicv-opencl) and the other in [test_videoio](https://pullrequest.opencv.org/buildbot/builders/precommit_windows64/builds/102981/steps/test_videoio) with

```[ERROR:0@149.540] global cap_ffmpeg_impl.hpp:1286 open VIDEOIO/FFMPEG: Failed to initialize VideoCapture```

Both of these seem to be unrelated, but let me know if I can check anything in the PR.

definitelyuncertain commented:
@asmorkalov

FYI, I updated the perf sanity data due to the change in coefficients for good measure, even though it was passing with the previous data.

definitelyuncertain commented:
@asmorkalov Just checking in to see if you're waiting on anything to get done. The builds are all passing so I'm assuming it's good to go.
